### PR TITLE
ACTION REQUESTED:  Please review, correct and approve the Merge to bring forward Marc's RocketChat teams work.

### PR DESCRIPTION
**Pull request by jmscraig**
CONTEXT:  This is a merge of the Original Rocket.Chat for teams commits written by Marc Enriquez with the go forward TeamTrials  "feature/teams" branch that was created from and is tracking the Sept 13th version of 'develop' branch on the original RocketChat repository (h--ps://github.com/RocketChat/Rocket.Chat).     

A.) ACTION REQUESTED:    This merge request brings forward the work Marc has done to create the Rocket.Chat features/teams into the go-forward TeamTrials Rocket.Chat Repository.    Please review the resolution choices for merge conflicts and validate that no significant code changes or additions are missing.    

B.) New Updated TeamTrials RocketChat Repository   - h--ps://github.com/TeamTrials/Rocket.Chat   
We have a new, clean TeamTrials / RocketChat Repository        
- Given our lack of dependency or risks on the intermediate changes upstream and desire to maintain simple, clean supportable, linage of the changes Marc made to create the 'teams' features I: 
- Deleted the existing Rocket.Chat fork replica in the TeamTrails org GitHub account
- Created a new fork of the original Rocket.Chat repository in the TeamTrials org GitHub account (our new "TeamTrials Rocket.Chat repository").

C.) We have Two primary TeamTrials RocketChat Teams (ttrc) branches to work with. 
1) Our ~"MASTER BRANCH" branch for our RocketChat work will be "feature/teams"  -  h--ps://github.com/TeamTrials/Rocket.Chat/tree/feature/teams 
- From the 'develop' branch in the original upstream RocketChat / RocketChat repository I created and set tracking for a new, go forward local branch named "features/teams" In the new TeamTrials Rocket.Chat repository. 
2) Our primary DEVELOPMENT BRANCH for our RocketChat work will be "feature/teams-develop"    h--ps://github.com/TeamTrials/Rocket.Chat/tree/feature/teams-develop   
-  Our new "feature/teams-develop" branch was created from and has tracking set to our new From our new ~master branch, "feature/teams" 

D.) Merge & Merge Conflict  Resolution Approach:   

Traditional Merge or Rebase to keep the repository we have using was not a good option.
- The TTRC feature/teams branch was 850+ commits behind the upstream 'develop' branch and with a 'diff' of changes to 1817 files and 116,000 lines of code.  Numerous merges & rebases had taken place on the upstream 'develop' since the last syncing with the TTRC features/team branch so a rebase or merge preserving time order of changes  was not possible nor would the chaos introduced to files we need to support desirable.

The merge to bring Marc's work forward
- Bypassing chaos of the 850+ commits in the upstream I merged the 'features/tream branch on Marc's repository into the new, go forward TeamTrials forward features/teams branch to produce a single, simple merge commit showing just the 35 files with the more significant changes from Marc.   Marc, please validate this list of 35 files did not miss any files we need to keep tracking.
- I executed an attempt to preserve 100% of the code and changes I believed to have come from Marc. 
- The changes I made were limited to 1) ESLinting (spaces, commas, etc..) to reduces chaos in this change review generated from near zero risk linting updates, and 2)  matching / leveling brackets and integrating edits when both Marc and the upstream rc core team were making changes to the same function.    
- I added comments at a few key areas to describe the changes or call out questions for Marc to answer.

Merge Conflicts:  
- The following 25 files had merge conflicts.
- The conflicts were resolved with the tool "VS Code".   Going forward the use of the popular, free VS Code will show the code change linage well during our development and support efforts.     
        both modified:   packages/rocketchat-api/server/v1/im.js
        both modified:   packages/rocketchat-apps/server/bridges/rooms.js
        both modified:   packages/rocketchat-error-handler/server/lib/RocketChat.ErrorHandler.js
        both modified:   packages/rocketchat-integrations/server/processWebhookMessage.js
        both modified:   packages/rocketchat-lib/client/lib/cachedCollection.js
        both modified:   packages/rocketchat-lib/client/lib/openRoom.js
        both modified:   packages/rocketchat-lib/lib/RoomTypesCommon.js
        both modified:   packages/rocketchat-lib/lib/roomTypes/direct.js
        both modified:   packages/rocketchat-lib/lib/roomTypes/private.js
        both modified:   packages/rocketchat-lib/lib/roomTypes/public.js
        both modified:   packages/rocketchat-lib/server/functions/createRoom.js
        both modified:   packages/rocketchat-lib/server/functions/getRoomByNameOrIdWithOptionToJoin.js
        both modified:   packages/rocketchat-lib/server/models/Messages.js
        both modified:   packages/rocketchat-lib/server/models/Rooms.js
        both modified:   packages/rocketchat-lib/server/models/Subscriptions.js
        both modified:   packages/rocketchat-livechat/.app/client/lib/fromApp/RoomHistoryManager.js
        both modified:   packages/rocketchat-ui-flextab/client/tabs/userActions.js
        both modified:   packages/rocketchat-ui-sidenav/client/listPrivateGroupsFlex.js
        both modified:   packages/rocketchat-ui-sidenav/client/roomList.js
        both modified:   packages/rocketchat-ui/client/lib/RoomHistoryManager.js
        both modified:   packages/rocketchat-ui/client/lib/RoomManager.js
        both modified:   server/methods/createDirectMessage.js
        both modified:   server/publications/room.js
        both modified:   server/publications/subscription.js
        both modified:   server/startup/initialData.js

E.) Marc's  (h--ps://github.com/auterium) original four Rocket.Chat feature/teams commits: 

4. Change roomId generation to prevent issues with jitsi 7572135
h--ps://github.com/RocketChat/Rocket.Chat/commit/7572135e1f1c60ee19d64d13f73cd2c03d0de675

3. Add name to room creation of direct messages to prevent issue ee97ee9
h--ps://github.com/RocketChat/Rocket.Chat/commit/ee97ee944c5682d77ff739b2de7b62cfe21d39db

2. Merge branch 'develop' into feature/teams    58c4e0a
h--ps://github.com/RocketChat/Rocket.Chat/commit/58c4e0a2d3757cd3c6887a6beb9d40befe4ba2d9

1. Add basic support for teams 7a0e697
h--ps://github.com/RocketChat/Rocket.Chat/commit/7a0e697215ee131c540c5191b204243a893e059c

### CODE REVIEW

File: packages/rocketchat-lib/server/functions/getRoomByNameOrIdWithOptionToJoin.js [hunk 1]
```diff
@@ -1,7 +1,11 @@
 /* globals RocketChat */
 import _ from 'underscore';
 
+<<<<<<< HEAD
```
Comment (auterium) [inline packages/rocketchat-lib/server/functions/getRoomByNameOrIdWithOptionToJoin.js]: Here's a merge conflict that didn't got resolved

File: packages/rocketchat-lib/server/models/Rooms.js [hunk 2]
```diff
@@ -5,9 +5,16 @@ class ModelRooms extends RocketChat.models._Base {
 	constructor(...args) {
 		super(...args);
 
-		this.tryEnsureIndex({ name: 1 }, { unique: 1, sparse: 1 });
+
 		this.tryEnsureIndex({ default: 1 });
 		this.tryEnsureIndex({ t: 1 });
+
+		// ttrc additions or edits from Marc
```
Comment (auterium) [inline packages/rocketchat-lib/server/models/Rooms.js]: Nonde of these indexes are necessary any more. Mine from line 13 was required as the unique index on "name" existed and was generating conflict on having multiple channels with the same name.

As for the "usernames" index, it should be removed too

File: packages/rocketchat-lib/server/models/Subscriptions.js [hunk 3]
```diff
@@ -21,6 +21,22 @@ class ModelSubscriptions extends RocketChat.models._Base {
 		this.tryEnsureIndex({ autoTranslate: 1 }, { sparse: 1 });
 		this.tryEnsureIndex({ autoTranslateLanguage: 1 }, { sparse: 1 });
 		this.tryEnsureIndex({ 'userHighlights.0': 1 }, { sparse: 1 });
+
+		// ttrc additions 
+		// ttrc - Marc, note for eslinting I removed quotes from name, t, code and team. please confirm.
```
Comment (auterium) [inline packages/rocketchat-lib/server/models/Subscriptions.js:26]: I'm fine with this

File: packages/rocketchat-lib/server/models/Subscriptions.js [hunk 4]
```diff
@@ -21,6 +21,22 @@ class ModelSubscriptions extends RocketChat.models._Base {
...
 		this.tryEnsureIndex({ 'userHighlights.0': 1 }, { sparse: 1 });
+
+		// ttrc additions 
+		// ttrc - Marc, note for eslinting I removed quotes from name, t, code and team. please confirm. 
+		this.tryEnsureIndex({ 'u._id': 1, name: 1, t: 1, code: 1, team: 1 }, { unique: 1 });
+
+		// ttrc - Marc, this was removed by rc core team.  please confirm we dont need it.
+		// this.tryEnsureIndex({ 'unread': 1 });
+
+		// ttrc - Marc, these have all been removed by rc core team. please confirm we need all rows.
```
Comment (auterium) [inline packages/rocketchat-lib/server/models/Subscriptions.js]: No, we don't. You can safely remove them

**Review by auterium — CHANGES_REQUESTED** [review]

File: packages/rocketchat-lib/server/models/Subscriptions.js (hunk 4 above)
Comment (jmscraig) [reply packages/rocketchat-lib/server/models/Subscriptions.js]: Removing this block

=============================

// ttrc - Marc, these have all been removed by rc core team. please confirm we need all rows.
this.cache.ensureIndex('rid', 'array');
this.cache.ensureIndex('u._id', 'array');
this.cache.ensureIndex('name', 'array');
this.cache.ensureIndex(['rid', 'u._id'], 'unique');

File: packages/rocketchat-lib/server/functions/getRoomByNameOrIdWithOptionToJoin.js (hunk 1 above)
Comment (jmscraig) [reply packages/rocketchat-lib/server/functions/getRoomByNameOrIdWithOptionToJoin.js]: Done

File: packages/rocketchat-lib/server/models/Rooms.js (hunk 2 above)
Comment (jmscraig) [reply packages/rocketchat-lib/server/models/Rooms.js]: Removed both lines and submitted a commit

File: packages/rocketchat-lib/server/models/Subscriptions.js (hunk 4 above)
Comment (jmscraig) [inline packages/rocketchat-lib/server/models/Subscriptions.js]: Marc, please review / comment on lines 38 & 39.  Do we still need this row?   
		// ttrc - this row was removed in Aug 2018 by rc core team to resolve a bug. 
		this.cache.ensureIndex(['name', 'team', 'u._id'], 'unique');

File: packages/rocketchat-lib/server/functions/getRoomByNameOrIdWithOptionToJoin.js [hunk 5]
```diff
@@ -37,7 +38,11 @@ RocketChat.getRoomByNameOrIdWithOptionToJoin = function _getRoomByNameOrIdWithOp
 			}
 
 			room = Meteor.runAsUser(currentUserId, function() {
+<<<<<<< HEAD
```
Comment (auterium) [inline packages/rocketchat-lib/server/functions/getRoomByNameOrIdWithOptionToJoin.js]: There's still a merge error here

**Review by auterium — CHANGES_REQUESTED** [review]

File: packages/rocketchat-lib/server/functions/getRoomByNameOrIdWithOptionToJoin.js (hunk 5 above)
Comment (jmscraig) [reply packages/rocketchat-lib/server/functions/getRoomByNameOrIdWithOptionToJoin.js]: Cleared.  Feels Like I have now  done this exact same commit twice but maybe the first time is really still open  in a browser somewhere.

Comment (jmscraig) [reply packages/rocketchat-lib/server/functions/getRoomByNameOrIdWithOptionToJoin.js]: There are now two commits to clear that merge conflict  
https://github.com/TeamTrials/Rocket.Chat/pull/2/commits/224dfac5618483c1ce2c2bee8857add978f5ce51

https://github.com/TeamTrials/Rocket.Chat/pull/2/commits/5fef91143d577314cdfc8d65cc285466a876eee8

Comment (auterium) [reply packages/rocketchat-lib/server/functions/getRoomByNameOrIdWithOptionToJoin.js]: Actualy, there were 2 like merge errors in this file, but the first time you only removed 1. This is all good now

**Review by auterium — APPROVED** [review]

File: packages/rocketchat-lib/server/functions/getRoomByNameOrIdWithOptionToJoin.js (hunk 1 above)
Comment (jmscraig) [reply packages/rocketchat-lib/server/functions/getRoomByNameOrIdWithOptionToJoin.js]: @auterium   -- you approved the commit to close this out an hour ago below so seems we could click this conversation as resolved as well  

5fef911  Merged Marc's 'teams' changes without adjustment to getRoomByNameOrId… "

**Comment by jmscraig** [comment]
@Picassotrigger @blueoceancruiser  Marc has reviewed, drove and closed on some improvements in the pull request.

Approval of the Pull Request is required by one of you two to complete the this "Risk High - Major Changes" Pull Request.   Please feel free to post any additional questions or topics you feel we should drive to resolution to complete the final scope and approval this pull request.

**Review by Picassotrigger — APPROVED** [review]

**Pull request merged**